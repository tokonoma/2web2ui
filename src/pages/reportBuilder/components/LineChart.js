import asyncComponent from 'src/components/asyncComponent/asyncComponent';

export default asyncComponent(() => import('./async/LineChart'));
