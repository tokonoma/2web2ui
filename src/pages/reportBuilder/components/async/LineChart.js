/*
  This component is meant to be loaded asynchronously, do not import directly.
  See ../LineChart.js for async export
*/

import React from 'react';
import {
  Bar,
  ComposedChart,
  Line,
  ReferenceLine,
  ResponsiveContainer,
  Tooltip,
  XAxis,
  YAxis,
} from 'recharts';
import moment from 'moment';
import styles from './LineChart.module.scss';
import { tokens } from '@sparkpost/design-tokens-hibana';
import { Box, Text } from 'src/components/matchbox';

const identity = a => a;

function orderDesc(a, b) {
  return b.value - a.value;
}

const CustomTooltip = ({ payload, label, labelFormatter, formatter }) => {
  return (
    <Box borderRadius="200" padding="200" bg="gray.1000">
      <Box fontSize="100" color="gray.200" mb="100">
        {labelFormatter(label)}
      </Box>
      {payload.map(entry => (
        <Box key={`report_chart_${entry.name}`} mb="100">
          <Box justifyContent="space-between" alignItems="center" display="flex">
            <Box display="inline-flex" alignItems="center">
              <Box //The color circle
                height="14px"
                width="14px"
                borderRadius="circle"
                mr={tokens.spacing_300}
                backgroundColor={entry.stroke}
              />
              <Text as="span" fontSize="100" color="white">
                {entry.name}
              </Text>
            </Box>
            <Box ml="800">
              <Text fontSize="100" textAlign="right" color="white">
                {formatter(entry.value)}
              </Text>
            </Box>
          </Box>
        </Box>
      ))}
    </Box>
  );
};

export default class SpLineChart extends React.Component {
  renderLines() {
    const { lines = [] } = this.props;
    return lines.map(line => {
      const lineProps = {
        strokeWidth: 2,
        animationDuration: 400,
        activeDot: false,
        dot: false,
        type: 'linear',
        ...line,
      };
      return <Line {...lineProps} />;
    });
  }

  renderReferenceLines() {
    const { referenceLines = [] } = this.props;
    return referenceLines.map(props => <ReferenceLine {...props} />);
  }

  // Manually generates X axis ticks
  getXTicks() {
    const { data, precision } = this.props;
    let ticks;

    // Shows ticks every Sunday
    if (precision === 'day' && data.length > 15) {
      ticks = data.reduce((acc, { ts }) => {
        if (moment(ts).isoWeekday() === 7) {
          acc.push(ts);
        }
        return acc;
      }, []);
    }

    // Show ticks every 15 minutes
    if (precision === '1min') {
      ticks = data.reduce((acc, { ts }) => {
        if (moment(ts).minutes() % 15 === 0) {
          acc.push(ts);
        }
        return acc;
      }, []);
    }

    // Show ticks every 30 minutes
    if (precision === '15min') {
      ticks = data.reduce((acc, { ts }) => {
        if (moment(ts).minutes() % 30 === 0) {
          acc.push(ts);
        }
        return acc;
      }, []);
    }

    return ticks;
  }

  render() {
    const {
      data,
      lines = [],
      syncId,
      xTickFormatter = identity,
      yTickFormatter = identity,
      yScale = 'linear', // eslint-disable-line
      tooltipLabelFormatter = identity,
      tooltipValueFormatter = identity,
      showXAxis,
      yLabel,
    } = this.props;

    return (
      <div className={styles.ChartWrapper}>
        <ResponsiveContainer width="99%" height={170 + 30 * lines.length}>
          <ComposedChart barCategoryGap="3%" syncId={syncId} data={data}>
            <XAxis
              axisLine={false}
              dataKey="ts"
              height={30}
              hide={!showXAxis}
              interval="preserveStartEnd"
              scale="auto"
              tickFormatter={xTickFormatter}
              tickLine={false}
              ticks={this.getXTicks()}
            />
            <YAxis
              axisLine={false}
              domain={['dataMin', 'dataMax']}
              interval="preserveStartEnd"
              padding={{ top: 8, bottom: 8 }}
              scale={yScale}
              tickFormatter={yTickFormatter}
              tickLine={false}
              width={60}
            />
            <Tooltip
              cursor={{ fill: '#ff0' }}
              content={<CustomTooltip />}
              isAnimationActive={false}
              itemSorter={orderDesc}
              labelFormatter={tooltipLabelFormatter}
              formatter={tooltipValueFormatter}
            />
            <Bar background={{ fill: tokens.color_gray_100 }} />
            {/* {this.renderReferenceLines()} */}
            {this.renderLines()}
          </ComposedChart>
        </ResponsiveContainer>
        <span className="sp-linechart-yLabel">{yLabel}</span>
      </div>
    );
  }
}
