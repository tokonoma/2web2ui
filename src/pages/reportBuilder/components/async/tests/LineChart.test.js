import React from 'react';
import LineChart from '../LineChart';
import { shallow } from 'enzyme';

describe('LineChart', () => {
  let props;
  let wrapper;

  beforeEach(() => {
    props = {
      syncId: 'summaryChart',
      precision: 'hour',
      showXAxis: true,
      yLabel: 'Count',
      yScale: 'linear',
      data: [
        {
          count_targeted: 10856,
          count_rendered: 5637,
          count_accepted: 9525,
          count_bounce: 1064,
          ts: '2018-11-06T10:00:00-05:00',
        },
        {
          count_targeted: 10914,
          count_rendered: 5628,
          count_accepted: 9428,
          count_bounce: 1143,
          ts: '2018-11-06T11:00:00-05:00',
        },
      ],
    };

    wrapper = shallow(<LineChart {...props} />);
  });

  it('should render', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
