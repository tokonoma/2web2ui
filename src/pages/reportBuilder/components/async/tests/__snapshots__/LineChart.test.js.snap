// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LineChart should render 1`] = `
<div
  className="ChartWrapper"
>
  <ResponsiveContainer
    debounce={0}
    height={170}
    width="99%"
  >
    <LineChart
      barCategoryGap="10%"
      barGap={4}
      data={
        Array [
          Object {
            "count_accepted": 9525,
            "count_bounce": 1064,
            "count_rendered": 5637,
            "count_targeted": 10856,
            "ts": "2018-11-06T10:00:00-05:00",
          },
          Object {
            "count_accepted": 9428,
            "count_bounce": 1143,
            "count_rendered": 5628,
            "count_targeted": 10914,
            "ts": "2018-11-06T11:00:00-05:00",
          },
        ]
      }
      layout="horizontal"
      margin={
        Object {
          "bottom": 5,
          "left": 5,
          "right": 5,
          "top": 5,
        }
      }
      reverseStackOrder={false}
      stackOffset="none"
      syncId="summaryChart"
    >
      <CartesianGrid
        fill="none"
        horizontal={true}
        horizontalFill={Array []}
        horizontalPoints={Array []}
        stroke="#ccc"
        strokeDasharray="4 1"
        vertical={false}
        verticalFill={Array []}
        verticalPoints={Array []}
      />
      <XAxis
        allowDataOverflow={false}
        allowDecimals={true}
        allowDuplicatedCategory={true}
        dataKey="ts"
        domain={
          Array [
            0,
            "auto",
          ]
        }
        height={30}
        hide={false}
        interval="preserveEnd"
        mirror={false}
        orientation="bottom"
        padding={
          Object {
            "left": 0,
            "right": 0,
          }
        }
        reversed={false}
        scale="auto"
        tickCount={5}
        tickFormatter={[Function]}
        type="category"
        width={0}
        xAxisId={0}
      />
      <YAxis
        allowDataOverflow={false}
        allowDecimals={true}
        allowDuplicatedCategory={true}
        domain={
          Array [
            "dataMin",
            "dataMax",
          ]
        }
        height={0}
        hide={false}
        interval="preserveStartEnd"
        mirror={false}
        orientation="left"
        padding={
          Object {
            "bottom": 8,
            "top": 8,
          }
        }
        reversed={false}
        scale="linear"
        tickCount={5}
        tickFormatter={[Function]}
        tickLine={false}
        type="number"
        width={60}
        yAxisId={0}
      />
      <Tooltip
        active={false}
        animationDuration={400}
        animationEasing="ease"
        contentStyle={Object {}}
        coordinate={
          Object {
            "x": 0,
            "y": 0,
          }
        }
        cursor={true}
        cursorStyle={Object {}}
        filterNull={true}
        formatter={[Function]}
        isAnimationActive={false}
        itemSorter={[Function]}
        itemStyle={Object {}}
        labelFormatter={[Function]}
        labelStyle={Object {}}
        offset={10}
        separator=" : "
        useTranslate3d={false}
        viewBox={
          Object {
            "x1": 0,
            "x2": 0,
            "y1": 0,
            "y2": 0,
          }
        }
        wrapperStyle={Object {}}
      />
    </LineChart>
  </ResponsiveContainer>
  <span
    className="sp-linechart-yLabel"
  >
    Count
  </span>
</div>
`;
